### PR TITLE
Optimize top_set_bit(::BigInt) and hash(::Real)

### PR DESCRIPTION
**Pull request by LilithHafner**
EDIT: this PR now does a fair bit more than #49986 did, see later comments.

Re-lands #49986, but computes `abs` first. This should be safe now because `ndigits0z(x, 2)` and `top_set_bit(abs(x))` are semantically equivalent whereas before I was falsely assuming `x > 0`.

The performance gain is now only 1.4x instead of 1.5x, which could be noise or the cost of `abs` (probably noise)

### CODE REVIEW

**Comment by LilithHafner** [comment]
Empirical check that `top_set_bit(abs(x)) == Base.ndigits0z(x, 2)` is valid:

```julia
for T in [UInt8, Int8] 
    for i in typemin(T):typemax(T)
        for j in (i, big(i), Int(i), unsigned(Int(i)))
            Base.top_set_bit(abs(j)) == Base.ndigits0z(j, 2) || error()
        end
    end
end
```

**Comment by oscardssmith** [comment]
While we're at it, is there a reason we're doing `hash(Int64(num) << Int(pow), h)` rather than `hash(Int64(x))`? (and similarly `hash(Float64(x))` instead of `hash(Int64(num) << Int(pow), h)`

**Comment by LilithHafner** [comment]
I suppose the existing version could be better for types that are slow to decompose and to convert to integers or floating-point values. I benchmarked no difference on any of
```julia
@btime hash(x) setup=(x=rand(Int)//1)
@btime hash(x) setup=(x=rand(Int)//1024)
@btime hash(x) setup=(x=rand(Int)//rand(Int))
@btime hash(x) setup=(x=rand(Int128))
```
for switching from `num << pow` to `x` in those three places, so the case to make that switch is pretty weak.

File: base/float.jl [hunk 1]
```diff
@@ -674,34 +674,27 @@ function hash(x::Real, h::UInt)
...
         if -1074 <= right
-            if 0 <= right && left <= 64
-                left <= 63                     && return hash(Int64(num) << Int(pow), h)
-                signbit(num) == signbit(den)   && return hash(UInt64(num) << Int(pow), h)
+            if 0 <= right
+                left <= 63 && return hash(Int64(num) << Int(den_z), h)
+                left <= 64 && !signbit(num) && return hash(UInt64(num) << Int(den_z), h)
             end # typemin(Int64) handled by Float64 case
-            left <= 1024 && left - right <= 53 && return hash(ldexp(Float64(num),pow), h)
+            left <= 1024 && left - right <= 53 && return hash(ldexp(Float64(num), den_z), h)
```
Comment (oscardssmith) [inline base/float.jl:690]: are we sure that with this change `num` will fit in a `Float64`?

**Comment by LilithHafner** [comment]
I reran all the benchmarks that have come up in this and #49998 and reported `before => after` all benchmarks are allocation free.
```julia
@btime Base.hash(x, UInt(1234)) setup=(x=Int128(rand(Int)));
  # 12.262 ns => 3.958 ns
@btime Base.hash(x, UInt(1234)) setup=(x=Int128(rand(Int128)));
  # 24.975 ns => 16.367 ns
@btime Base.top_set_bit(abs(x)) setup=(x=big(rand(Int)));
  # 11.469 ns => 2.541 ns
@btime Base.top_set_bit(abs(x)) setup=(x=big(rand(Int128)));
  # 11.512 ns => 2.583 ns
@btime hash(x) setup=(x=big(rand(Int))//big(1));
  # 33.442 ns => 24.807 ns
@btime hash(x) setup=(x=big(rand(Int128))//big(1));
  # 41.625 ns => 35.247 ns
@btime hash(x) setup=(x=rand(Int)//1);
  # 3.708 ns  => 3.750 ns
@btime hash(x) setup=(x=rand(Int)//1024);
  # 3.837 ns  => 3.833 ns
@btime hash(x) setup=(x=rand(Int)//rand(Int));
  # 14.780 ns => 14.780 ns
@btime hash(x) setup=(x=rand(Int128));
  # 23.845 ns => 16.784 ns
```

**Comment by oscardssmith** [comment]
To be clear: are those benchmarks including #49998?

**Comment by LilithHafner** [comment]
Those results are just this PR.

**Comment by oscardssmith** [comment]
interesting. That somehow seems to be faster than https://github.com/JuliaLang/julia/pull/49998 which is rather surprising to me. I guess our compiler is doing a really good job cleaning this up for `Int128`.

**Comment by LilithHafner** [comment]
Yes. It took a fair bit of fiddling before it matched the performance of #49998, but once it did I knew it was pretty good :). I expect the reason this appears more impactful than #49998 is that the impact is more pronounced on my machine than yours.

**Comment by LilithHafner** [comment]
Oops, I hadn't noticed :)

**Comment by oscardssmith** [comment]
I think this is good to go, but I do kind of want a review from @vtjnash here.

**Comment by oscardssmith** [comment]
Closing since https://github.com/JuliaLang/julia/pull/49996 gets the same performance.

**Comment by oscardssmith** [comment]
(oops, closed the wrong one)

**Pull request merged**